### ISSUE TITLE
False positive on assigning-non-slot when following standard idiom for __deepcopy__ in a class hierarchy

### ISSUE THREAD

**Issue by fizbin** [PyCQA/pylint#4639]
### Steps to reproduce

Run `pylint` version 2.9.0 (or the current pre-release) on this file:

```py
"""
A demonstration of pylint being excessive with assigning-non-slot in pylint 2.9.0
"""


class ParentClass:
    """
    Parent class for demonstration
    """

    __slots__ = ("parent_slot",)

    def __init__(self):
        self.parent_slot = 0

    def __deepcopy__(self, memo):
        cls = self.__class__
        copied = cls.__new__(cls)
        copied.parent_slot = self.parent_slot
        return copied

    def twiddle_parent_slot(self):
        """
        Here mainly to avoid a "too few public methods" warning
        """
        self.parent_slot += 1


class ChildClass(ParentClass):
    """
    Child class for demonstration
    """

    __slots__ = ("child_slot",)

    def __init__(self):
        super().__init__()
        self.child_slot = ""

    def __deepcopy__(self, memo):
        copied: ChildClass = super().__deepcopy__(memo)
        copied.child_slot = self.child_slot
        return copied
```

### Current behavior

```
************* Module test
test.py:42:8: E0237: Assigning to attribute 'child_slot' not defined in class slots (assigning-non-slot)

------------------------------------------------------------------
Your code has been rated at 7.50/10 (previous run: 7.50/10, +0.00)
```

### Expected behavior

(this is the output with pylint 2.8.3)

```

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

```

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0-a4
astroid 2.6.2
Python 3.8.3 (v3.8.3:6f8c8320e9, May 13 2020, 16:29:34) 
[Clang 6.0 (clang-600.0.57)]
```

**Comment by jacobtylerwalls** [comment]
Thanks for the report. Although a duplicate of https://github.com/PyCQA/pylint/issues/981 and https://github.com/PyCQA/astroid/issues/457 I'll link this and mention the test case.
